scripts/portal-fetcher/openshift-docs-downloader.py

```diff
@@ -0,0 +1,1186 @@
+
+def construct_base_url(version, specific_doc=None):
+    """
+    Construct the base URL for OpenShift documentation
+    
+    Args:
+        version (str): OpenShift version
+        specific_doc (str): Optional specific document to download
+        
+    Returns:
+        tuple: (base_url, output_dir)
+    """
+    if specific_doc:
+        # If a specific document is requested, target only that document
+        base_url = f"https://docs.redhat.com/en/documentation/openshift_container_platform/{version}/html-single/{specific_doc}"
+        output_dir = f"./openshift-docs/{version}/html-single/{specific_doc}"
+        
+        # Check if index.html is explicitly specified, if not, add it
+        if not base_url.endswith("index"):
+            if not base_url.endswith("/"):
+                base_url = f"{base_url}/"
+    else:
+        # Otherwise, target the entire documentation
+        base_url = f"https://docs.redhat.com/en/documentation/openshift_container_platform/{version}"
+        output_dir = f"./openshift-docs/{version}"
+    
+    return (base_url, output_dir)
+
+async def main_async(args):
+    """
+    Main async function to run the downloader
+    
+    Args:
+        args: Command line arguments
+    """
+    # Apply rate limiting if requested
+    apply_rate_limiting(args.rate_limit)
+    
+    # Construct base URL and output directory
+    base_url, output_dir = construct_base_url(args.version, args.specific_doc)
+    
+    if args.output_dir:
+        if args.specific_doc:
+            output_dir = f"{args.output_dir}/{args.version}/html-single/{args.specific_doc}"
+        else:
+            output_dir = f"{args.output_dir}/{args.version}"
+    
+    logger.info(f"Base URL: {base_url}")
+    logger.info(f"Output directory: {output_dir}")
+    
+    # Run the downloader
+    verification_passed, toc_verification_passed, elapsed_time = await run_downloader(
+        base_url, 
+        output_dir, 
+        concurrency=args.concurrency, 
+        force=args.force, 
+        skip_toc=args.skip_toc_verification
+    )
+    
+    return verification_passed and toc_verification_passed
+
+def main():
+    """Main entry point"""
+    parser = argparse.ArgumentParser(description="Download Red Hat OpenShift Documentation")
+    parser.add_argument(
+        "--version", 
+        type=str, 
+        default="4.18",
+        help="OpenShift version (default: 4.18)"
+    )
+    parser.add_argument(
+        "--output-dir", 
+        type=str, 
+        default="./openshift-docs",
+        help="Output directory (default: ./openshift-docs)"
+    )
+    parser.add_argument(
+        "--concurrency", 
+        type=int, 
+        default=5,
+        help="Number of concurrent downloads (default: 5)"
+    )
+    parser.add_argument(
+        "--force",
+        action="store_true",
+        help="Force download of all files even if not modified"
+    )
+    parser.add_argument(
+        "--rate-limit",
+        type=float,
```

Comment:
I'm not sure if fractional seconds are useful, we could go with int. And attach range validation to the command line parser, like so
```
def check_positive_int(value):
    ivalue = int(value)
    if ivalue < 0:
        raise argparse.ArgumentTypeError(f"{value} is not a non-negative integer")
    return ivalue

parser = argparse.ArgumentParser()
parser.add_argument('--count', type=check_positive_int)
```
And then remove the range check from `apply_rate_limiting()`. Basically, if the script is running, all the command-line parameters are of correct types and valid ranges.